File: sdks/typescript/test/flexemarkets-live-server.test.ts
/**
 * Live-server smoke test for the TypeScript SDK. Mirrors the Java
 * FlexemarketsLiveServerTest. Catches catastrophic protocol regressions
 * (V0/V1 endpoint path drift, HATEOAS envelope shape, WS subscribe
 * destination, etc.) without needing a heavy test-server harness.
 *
 * Opt-in via two preconditions, both checked by `liveServerReady`:
 *   1. ~/.fm/credential and ~/.fm/endpoint both exist
 *   2. FM_LIVE_TESTS=1 env var is set
 *
 * Without those the suite self-skips so CI doesn't false-fail on a
 * missing server.
 *
 * Read-only: picks the endpoint's marketplace, opens a Desk,
 * verifies the accessors don't throw, closes. No orders submitted; no
 * marketplaces created. Idempotent.
 */

import { existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { test } from "node:test";
import assert from "node:assert/strict";

import { Flexemarkets } from "../src/index.ts";

function liveServerReady(): boolean {
  if (process.env.FM_LIVE_TESTS !== "1") return false;
  const home = homedir();
  return existsSync(join(home, ".fm", "credential")) && existsSync(join(home, ".fm", "endpoint"));
}

const live = liveServerReady() ? test : test.skip;

live("connects to local server and lists marketplaces", async () => {
  const fm = await Flexemarkets.connect(null, null, "fm-sdk-live-test/connectAndList");
  try {
    const marketplaces = await fm.marketplaces();
    assert.ok(marketplaces.length > 0, "expected at least the endpoint's marketplace");
  } finally {
    fm.close();
  }
});

live("opens a desk on the endpoint marketplace without throwing", async () => {
  const fm = await Flexemarkets.connect(null, null, "fm-sdk-live-test/desk");
  try {
    const marketplaceId = fm.endpointMarketplaceId;
    const desk = await fm.desk(marketplaceId);
    try {
      assert.equal(desk.marketplaceId, marketplaceId);
      assert.ok(desk.markets.length > 0, "marketplace should have at least one market");
      for (const m of desk.markets) {
        const book = desk.book(m.id);
        assert.ok(book !== null, `order book for market ${m.id} should be non-null`);
        // bestBuyPrice / bestSellPrice return -1 when empty; either
        // way they shouldn't throw, which is the real smoke-test
        // signal.
        assert.doesNotThrow(() => book!.bestBuyPrice());
        assert.doesNotThrow(() => book!.bestSellPrice());
      }
    } finally {
      desk.close();
    }
  } finally {
    fm.close();
  }
});

live("shares one desk across repeat calls for same marketplace", async () => {
  const fm = await Flexemarkets.connect(null, null, "fm-sdk-live-test/sharedDesk");
  try {
    const marketplaceId = fm.endpointMarketplaceId;
    const a = await fm.desk(marketplaceId);
    const b = await fm.desk(marketplaceId);
    try {
      assert.equal(a.marketplaceId, b.marketplaceId);
      assert.deepEqual(a.markets, b.markets);
      // Closing one handle must NOT close the shared desk — `b` should
      // still be usable.
      a.close();
      assert.doesNotThrow(() => b.markets);
    } finally {
      b.close();
    }
  } finally {
    fm.close();
  }
});

live("closing Flexemarkets force-closes any remaining shared desks", async () => {
  const fm = await Flexemarkets.connect(null, null, "fm-sdk-live-test/forceClose");
  const marketplaceId = fm.endpointMarketplaceId;
  // Intentionally don't close the handle — fm.close() must sweep it
  // up so the WS subscription is released.
  await fm.desk(marketplaceId);
  assert.doesNotThrow(() => fm.close());
});

File: sdks/java/fm-sdk/src/test/java/fm/FlexemarketsLiveServerTest.java
package fm;

import fm.model.Market;
import fm.model.Marketplace;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIf;


import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatNoException;

/**
 * Live-server smoke test for the Java SDK. Catches catastrophic protocol
 * regressions (V0/V1 endpoint path drift, snapshot header format, WS
 * subscribe destination, etc.) without needing the heavy in-process
 * Spring Boot setup the studies use.
 *
 * <p>Opt-in via two preconditions, both checked by {@link #liveServerReady()}:
 * <ol>
 *   <li>{@code ~/.fm/credential} and {@code ~/.fm/endpoint} both exist —
 *       the same config layout the SDK examples use.</li>
 *   <li>The {@code FM_LIVE_TESTS=1} environment variable is set. Without
 *       it the tests self-skip so CI doesn't false-fail on a missing
 *       server.</li>
 * </ol>
 *
 * <p>The smoke test deliberately stays read-only: it picks the
 * endpoint's marketplace, opens a {@link Desk}, verifies the
 * accessors don't throw, and closes. No orders submitted; no
 * marketplaces created. That keeps the test cheap and idempotent —
 * running it 100 times in a row leaves zero residue on the server.
 */
public class FlexemarketsLiveServerTest {

    private static Path _credentialPath;
    private static Path _endpointPath;

    @BeforeAll
    static void resolveConfig() {
        var home = System.getProperty("user.home");
        _credentialPath = Paths.get(home, ".fm", "credential");
        _endpointPath = Paths.get(home, ".fm", "endpoint");
    }

    /** Precondition gate for every test in this class. Returns true only
     *  when both config files exist AND {@code FM_LIVE_TESTS=1} is set
     *  in the environment. Without that env var the tests skip silently
     *  so CI doesn't false-fail. */
    static boolean liveServerReady() {
        if (!"1".equals(System.getenv("FM_LIVE_TESTS"))) return false;
        var home = System.getProperty("user.home");
        return Files.exists(Paths.get(home, ".fm", "credential"))
            && Files.exists(Paths.get(home, ".fm", "endpoint"));
    }

    @Test
    @EnabledIf("liveServerReady")
    void connectsToLocalServerAndListsMarketplaces() throws Exception {
        try (var fm = _connect("fm-sdk-live-test/connectAndList")) {
            List<Marketplace> marketplaces = fm.marketplaces();
            // The endpoint is configured to a specific marketplace, so
            // marketplaces() returns at least that one even if the
            // account has nothing else.
            assertThat(marketplaces).isNotEmpty();
        }
    }

    @Test
    @EnabledIf("liveServerReady")
    void opensADeskOnTheEndpointMarketplaceWithoutThrowing() throws Exception {
        try (var fm = _connect("fm-sdk-live-test/desk")) {
            long marketplaceId = fm.endpointMarketplaceId();
            // Use try-with-resources for the handle so the refcount
            // unwinds even if an assertion fails mid-test.
            try (Desk desk = (Desk) fm.desk(marketplaceId)) {
                assertThat(desk.marketplaceId()).isEqualTo(marketplaceId);
                assertThat(desk.markets()).isNotEmpty();
                for (Market m : desk.markets()) {
                    Book book = desk.book(m.id());
                    assertThat(book).isNotNull();
                    // bestBuyPrice / bestSellPrice return -1 when empty;
                    // either way they shouldn't throw, which is the real
                    // smoke-test signal.
                    assertThatNoException().isThrownBy(book::bestBuyPrice);
                    assertThatNoException().isThrownBy(book::bestSellPrice);
                }
            }
        }
    }

    @Test
    @EnabledIf("liveServerReady")
    void sharesOneDeskAcrossRepeatCallsForSameMarketplace() throws Exception {
        try (var fm = _connect("fm-sdk-live-test/sharedDesk")) {
            long marketplaceId = fm.endpointMarketplaceId();
            Desk a = fm.desk(marketplaceId);
            Desk b = fm.desk(marketplaceId);
            try {
                // Both handles see the same marketplace + market list —
                // the contract isn't that they're equal references, it's
                // that they expose the same state coherently.
                assertThat(a.marketplaceId()).isEqualTo(b.marketplaceId());
                assertThat(a.markets()).usingRecursiveComparison().isEqualTo(b.markets());
                // Closing one handle must NOT close the shared desk —
                // 'b' should still be usable.
                a.close();
                assertThatNoException().isThrownBy(b::markets);
            } finally {
                b.close();
            }
        }
    }

    @Test
    @EnabledIf("liveServerReady")
    void closingFlexemarketsForceClosesRemainingShared() throws Exception {
        Flexemarkets fm = _connect("fm-sdk-live-test/forceClose");
        long marketplaceId = fm.endpointMarketplaceId();
        Desk desk = fm.desk(marketplaceId);
        // Intentionally don't close 'desk' — Flexemarkets.close() must
        // sweep it up so the WS subscription is released.
        assertThatNoException().isThrownBy(fm::close);
    }

    private static Flexemarkets _connect(String clientDescription) throws IOException {
        // Flexemarkets.connect expects file paths (or a token / URL),
        // not the file contents. Same convention the ticker example uses.
        return Flexemarkets.connect(
                _credentialPath.toString(),
                _endpointPath.toString(),
                clientDescription);
    }
}
